src/sentry/interfaces/base.py

```diff
@@ -120,7 +124,62 @@ def __setattr__(self, name, value):
 
     @classmethod
     def to_python(cls, data):
-        return cls(**data)
+        """Creates a python interface object from the given raw data.
+
+        This function can assume fully normalized and valid data. It can create
+        defaults where data is missing but does not need to handle interface
+        validation.
+        """
+        return data and cls(**data)
+
+    @classmethod
+    def _normalize(cls, data, meta):
+        """Custom interface normalization. ``data`` is guaranteed to be a
+        non-empty mapping. Return ``None`` for invalid data.
+        """
+        return cls.to_python(data).to_json()
+
+    @classmethod
+    def normalize(cls, data, meta):
+        """Normalizes the given raw data removing or replacing all invalid
+        attributes. If the interface is unprocessable, ``None`` is returned
+        instead.
+
+        Errors are written to the ``meta`` container. Use ``Meta.enter(key)`` to
+        obtain an instance.
+
+        TEMPORARY: The transitional default behavior is to call to_python and
+        catch exceptions into meta data. To migrate, override ``_normalize``.
+        """
+
+        # Gracefully skip empty data. We treat ``None`` and empty objects the
+        # same as missing data. If there are meta errors attached already, they
+        # will remain in meta.
+        if not data:
+            return None
+
+        # Interface data is required to be a JSON object. Places where the
+        # protocol permits lists must be casted to a values wrapper first.
+        if not isinstance(data, Mapping):
+            meta.add_error('expected %s' % (cls.__name__,), data)
+            return None
+
+        try:
+            data = cls._normalize(data, meta=meta)
+        except Exception as e:
+            # XXX: InterfaceValidationErrors can be thrown in the transitional
+            # phase while to_python is being used for normalization. All other
+            # exceptions indicate a programming error and need to be reported.
+            if not isinstance(e, InterfaceValidationError):
+                logger.error('Discarded invalid value for interface: %s (%r)',
+                             cls.path, data, exc_info=True)
+
+            meta.add_error(e, value=data)
+            return None
+
+        # As with input data, empty interface data is coerced to None after
+        # normalization.
+        return data or None
```

Comment:
```suggestion
        # Otherwise, we apply ``to_json`` which has custom logic for pruning
        # empty keys.
        return cls.to_python(data).to_json() if data else None
```

@untitaker This is what I meant with Option 2 of my previous comment. We'd have a redundant pass through python, but could easily reuse `to_json` that way. Otherwise, we would have to figure out how to deduplicate `to_json` somehow.

Reply:
I don't understand why you need this. Shouldn't `_normalize` take care of pruning empty keys too? Custom implementors could even call `super()._normalize` if they want to.

Reply:
Mostly because I don't want duplicate code for pruning the same keys (`xyz or None`) in both `_normalize` and `to_json`.